### PR TITLE
Apigee instance

### PR DESCRIPTION
**Pull request by drebes**
If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
google_apigee_instance
```

### CODE REVIEW

File: mmv1/products/apigee/terraform.yaml [hunk 1]
```diff
@@ -47,11 +47,47 @@ overrides: !ruby/object:Overrides::ResourceOverrides
...
+    - !ruby/object:Provider::Terraform::Examples
+      # This is a more verbose version of the above that creates all
+      # the resources needed for the acceptance test.
+      name: "apigee_instance_basic_test"
+      primary_resource_id: "apigee_instance"
+      test_env_vars:
+        org_id: :ORG_ID
+        billing_account: :BILLING_ACCT
+      skip_docs: true
+      # Resource creation race
```
Comment (melinath) [inline mmv1/products/apigee/terraform.yaml:69]: Could you elaborate on this? If there is a resource creation race, I think that shouldn't impact VCR tests, since they don't run off real API requests.

File: mmv1/templates/terraform/examples/apigee_organization_cloud_full.tf.erb [hunk 2]
```diff
@@ -53,7 +53,7 @@ resource "google_apigee_organization" "org" {
   description                          = "Terraform-provisioned Apigee Org."
   project_id                           = data.google_client_config.current.project
   authorized_network                   = google_compute_network.apigee_network.id
-  runtime_database_encryption_key_name = google_kms_key.apigee_key.id
+  runtime_database_encryption_key_name = google_kms_crypto_key.apigee_key.id
```
Comment (melinath) [inline mmv1/templates/terraform/examples/apigee_organization_cloud_full.tf.erb:56]: This seems unrelated - just wanted to double-check it was included intentionally?

File: mmv1/products/apigee/terraform.yaml [hunk 3]
```diff
@@ -47,11 +47,47 @@ overrides: !ruby/object:Overrides::ResourceOverrides
...
+    autogen_async: true
+    import_format: ["{{org_id}}/instances/{{name}}", "{{org_id}}/{{name}}"]
+    examples:
+    - !ruby/object:Provider::Terraform::Examples
+      name: "apigee_instance_basic"
+      skip_test: true
+    - !ruby/object:Provider::Terraform::Examples
+      # This is a more verbose version of the above that creates all
+      # the resources needed for the acceptance test.
+      name: "apigee_instance_basic_test"
```
Comment (melinath) [inline mmv1/products/apigee/terraform.yaml:63]: Having separate `*_basic` and `*_basic_test` examples seems a bit odd but is in line with what the other apigee resources do. Do you happen to know the reasoning behind this split?

**Review by melinath — COMMENTED** [review]
In general this looks good to me. I have a couple of questions & I need to run the tests manually since the automatic run failed :-)

File: mmv1/products/apigee/terraform.yaml (hunk 1 above)
Comment (melinath) [reply mmv1/products/apigee/terraform.yaml:69]: It looks like this also prevents the new tests from getting run on this PR so I would definitely prefer to not mark these as skip_vcr unless absolutely necessary.

File: mmv1/products/apigee/terraform.yaml (hunk 3 above)
Comment (drebes) [reply mmv1/products/apigee/terraform.yaml:63]: The reason is that for the tests only, we need to create the org in its own project (since there can only be one org per project), although this is irrelevant in terms of documentation. We don't want to confuse the user that they need a new project to create the instance, but we also don't want to create the org on the acceptance test project as that may cause multiple simultaneous acceptance tests run to compete creating the (only) org for that project.

File: mmv1/templates/terraform/examples/apigee_organization_cloud_full.tf.erb (hunk 2 above)
Comment (drebes) [reply mmv1/templates/terraform/examples/apigee_organization_cloud_full.tf.erb:56]: Correct, unrelated, but a typo I found as I was writing the full test for instance. I could move this into its own PR but since it was doc only I though it was safe to include here.

File: mmv1/products/apigee/terraform.yaml (hunk 1 above)
Comment (drebes) [reply mmv1/products/apigee/terraform.yaml:69]: I just copied over this from the org tests (since this includes the org tests). It was introduced [here](https://github.com/GoogleCloudPlatform/magic-modules/pull/4401/files#diff-d03f223d530588d47ff6aa088a55e79abf68090146cb8954fbf036a8381a40daR30), perhaps @slevenick can comment on the reasoning behind it?

**Comment by drebes** [comment]
I'm wondering if we should work with the Apigee team to have an Apigee Org pre-boostrapped on the testing infra so we don't need to spawn a new one (in a new project) for every new Apigee resource we support? If you think so I can facilitate the conversation with them.

File: mmv1/templates/terraform/examples/apigee_organization_cloud_full.tf.erb (hunk 2 above)
Comment (melinath) [reply mmv1/templates/terraform/examples/apigee_organization_cloud_full.tf.erb:56]: I'm fine with keeping it here. :-)

**Comment by melinath** [comment]
@drebes I checked with the team and we'd be down with having something set up for the testing infra. :-) It looks like we wouldn't necessarily need to coordinate with the Apigee team; we have a system for "bootstrapping" infra that will be reused between test runs: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/third_party/terraform/utils/bootstrap_utils_test.go

Would something like that seem reasonable to you? If so, we could open a terraform-provider-google ticket to track the work (you would be welcome to contribute that work if you're interested, but it would probably make sense to track and work as a separate issue from this PR).

File: mmv1/products/apigee/terraform.yaml (hunk 1 above)
Comment (drebes) [reply mmv1/products/apigee/terraform.yaml]: I removed it for the time being until we hear back from @slevenick if what was the reason for introducing it for Orgs (or use a boostrapped Org instead).

**Comment by melinath** [comment]
@drebes heads up, now that the tests are running they're failing with the following error:

```
Error: Error when reading or editing Organization: googleapi: Error 403: Permission denied on resource project #147010595648.
              Details:
              [
                {
                  "@type": "type.googleapis.com/google.rpc.Help",
                  "links": [
                    {
                      "description": "Google developer console API key",
                      "url": "https://console.developers.google.com/project/147010595648/apiui/credential"
                    }
                  ]
                },
                {
                  "@type": "type.googleapis.com/google.rpc.ErrorInfo",
                  "domain": "googleapis.com",
                  "metadata": {
                    "consumer": "projects/147010595648",
                    "service": "apigee.googleapis.com"
                  },
                  "reason": "CONSUMER_INVALID"
                }
              ]
```

File: mmv1/products/apigee/terraform.yaml (hunk 1 above)
Comment (slevenick) [reply mmv1/products/apigee/terraform.yaml]: You will need to add it back.

In that test we enable multiple `google_project_service` in a single terraform apply. These are not actually individual resources but sub-resources on a single actual resource. The problem is that Terraform does not apply these in a deterministic order, so on one run it could add service 1 first, then 2 then 3. Recorded requests will then only work if the resources are created in that order. If another run in VCR mode tries to create 2 then 3 then 1 it will fail with an inconsistent provider message

Comment (melinath) [reply mmv1/products/apigee/terraform.yaml]: That makes sense - thanks, Sam!

**Comment by melinath** [comment]
Manual run of just the apigee instance test for the GA provider succeeded: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloud_ProviderGoogleCloudMmUpstream/174193

File: mmv1/products/apigee/terraform.yaml (hunk 1 above)
Comment (drebes) [reply mmv1/products/apigee/terraform.yaml:69]: I added skip_vcr back.

**Review by melinath — APPROVED** [review]
Per discussion with @priya-saxena21 and Dan, it looks like these tests are flakey rather than consistently failing. I'm approving for merge & we can look at making the tests more reliable separately.

**Comment by melinath** [comment]
Also, checked with Apigee product team during that discussion; their take was that creating a more permanent Organization would not help, since it is a quick and reliable step; the part that is failing is the Instance creation.

**Comment by drebes** [comment]
Ok, I was also thinking about creating a bootstrapped instance, since we'll have other resources under Instance coming up.

**Comment by melinath** [comment]
@drebes if that's useful for y'all definitely go for it! The 30+ min spin-up/spin-down is kinda brutal 😬  I think I misunderstood what you were thinking about doing :-)

**Comment by melinath** [comment]
As long as multiple test runs can happen simultaneously it shouldn't be an issue.
